File: backend/internal/rpc/polymarket.go
package rpc

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/timestamppb"
	polymarketv1 "selfquant/backend/gen/polymarket/v1"
	"selfquant/backend/internal/polymarket"
)

type PolymarketServer struct {
	polymarketv1.UnimplementedPolymarketServiceServer
	service   *polymarket.Service
	snapshots *polymarket.SnapshotStore
}

func NewPolymarketServer(
	service *polymarket.Service,
	snapshots *polymarket.SnapshotStore,
) *PolymarketServer {
	return &PolymarketServer{service: service, snapshots: snapshots}
}

func (s *PolymarketServer) ListMarkets(
	_ context.Context,
	request *polymarketv1.ListMarketsRequest,
) (*polymarketv1.ListMarketsResponse, error) {
	markets, version := s.snapshots.ListMarkets(
		request.GetAsset(), request.GetPeriod(), request.GetActiveOnly(),
	)
	items := make([]*polymarketv1.Market, 0, len(markets))
	for _, market := range markets {
		items = append(items, marketToProto(market))
	}
	return &polymarketv1.ListMarketsResponse{
		Items: items, ServerTime: timestamppb.Now(),
		SnapshotVersion: fmt.Sprintf("%d", version),
	}, nil
}

func (s *PolymarketServer) GetMarketSnapshot(
	_ context.Context,
	request *polymarketv1.GetMarketSnapshotRequest,
) (*polymarketv1.GetMarketSnapshotResponse, error) {
	snapshot, ok := s.snapshots.Get(request.GetMarketId())
	if !ok {
		return nil, status.Error(codes.NotFound, "market not found")
	}
	return &polymarketv1.GetMarketSnapshotResponse{
		Snapshot:   snapshotToProto(snapshot, snapshot.Series, true),
		ServerTime: timestamppb.Now(),
	}, nil
}

func (s *PolymarketServer) StreamMarketSnapshots(
	request *polymarketv1.StreamMarketSnapshotsRequest,
	stream polymarketv1.PolymarketService_StreamMarketSnapshotsServer,
) error {
	marketID := request.GetMarketId()
	initial, ok := s.snapshots.Get(marketID)
	if !ok {
		return status.Error(codes.NotFound, "market not found")
	}
	fullSeries := polymarket.DownsampleSeries(initial.Series, polymarket.MaxChartPoints)
	if err := stream.Send(&polymarketv1.StreamMarketSnapshotsResponse{
		Snapshot:   snapshotToProto(initial, fullSeries, true),
		ServerTime: timestamppb.Now(),
	}); err != nil {
		return err
	}
	updates, cancel := s.snapshots.Subscribe(marketID)
	defer cancel()
	heartbeat := time.NewTicker(15 * time.Second)
	defer heartbeat.Stop()
	for {
		select {
		case <-stream.Context().Done():
			return stream.Context().Err()
		case event := <-updates:
			proto := streamSnapshotEvent(event)
			if err := stream.Send(&polymarketv1.StreamMarketSnapshotsResponse{
				Snapshot: proto, ServerTime: timestamppb.Now(),
			}); err != nil {
				return err
			}
		case <-heartbeat.C:
			if err := stream.Send(&polymarketv1.StreamMarketSnapshotsResponse{
				ServerTime: timestamppb.Now(),
			}); err != nil {
				return err
			}
		}
	}
}

func streamSnapshotEvent(event polymarket.SnapshotEvent) *polymarketv1.MarketSnapshot {
	switch event.Kind {
	case polymarket.SnapshotEventFull:
		return snapshotToProto(
			event.Snapshot,
			polymarket.DownsampleSeries(event.Snapshot.Series, polymarket.MaxChartPoints),
			true,
		)
	case polymarket.SnapshotEventPriceDelta:
		return snapshotToProto(event.Snapshot, event.DeltaPoints, true)
	default:
		return snapshotToProto(event.Snapshot, nil, false)
	}
}

func (s *PolymarketServer) GetAccountSummary(
	ctx context.Context,
	request *polymarketv1.GetAccountSummaryRequest,
) (*polymarketv1.GetAccountSummaryResponse, error) {
	summary, err := s.service.GetAccountSummary(
		ctx, request.GetToken(), request.GetTradingAccountId(),
	)
	if err != nil {
		return nil, mapPolymarketError(err)
	}
	return &polymarketv1.GetAccountSummaryResponse{
		Summary: &polymarketv1.AccountSummary{
			TradingAccountId: summary.TradingAccountID,
			AccountName:      summary.AccountName, WalletAddress: summary.WalletAddress,
			AvailableBalance: summary.AvailableBalance,
			PositionValue:    summary.PositionValue, TotalAssets: summary.TotalAssets,
			SourceUpdatedAt: timestamppb.New(summary.SourceUpdatedAt), Stale: summary.Stale,
		},
	}, nil
}

func (s *PolymarketServer) ListPositions(
	ctx context.Context,
	request *polymarketv1.ListPositionsRequest,
) (*polymarketv1.ListPositionsResponse, error) {
	positions, stale, err := s.service.ListPositions(
		ctx, request.GetToken(), request.GetTradingAccountId(), false,
	)
	if err != nil {
		return nil, mapPolymarketError(err)
	}
	items := make([]*polymarketv1.Position, 0, len(positions))
	for _, position := range positions {
		items = append(items, positionToProto(position))
	}
	return &polymarketv1.ListPositionsResponse{
		Items: items, ServerTime: timestamppb.Now(), Stale: stale,
	}, nil
}

func (s *PolymarketServer) ListOpenOrders(
	ctx context.Context,
	request *polymarketv1.ListOpenOrdersRequest,
) (*polymarketv1.ListOpenOrdersResponse, error) {
	orders, stale, err := s.service.ListOpenOrders(
		ctx, request.GetToken(), request.GetTradingAccountId(), false,
	)
	if err != nil {
		return nil, mapPolymarketError(err)
	}
	items := make([]*polymarketv1.OpenOrder, 0, len(orders))
	for _, order := range orders {
		items = append(items, openOrderToProto(order))
	}
	return &polymarketv1.ListOpenOrdersResponse{
		Items: items, ServerTime: timestamppb.Now(), Stale: stale,
	}, nil
}

func (s *PolymarketServer) PlaceOrder(
	ctx context.Context,
	request *polymarketv1.PlaceOrderRequest,
) (*polymarketv1.PlaceOrderResponse, error) {
	order, err := s.service.PlaceOrder(ctx, polymarket.PlaceOrderInput{
		Token: request.GetToken(), TradingAccountID: request.GetTradingAccountId(),
		MarketID: request.GetMarketId(), Outcome: request.GetOutcome(),
		Side: request.GetSide(), Amount: request.GetAmount(),
		AmountUnit: request.GetAmountUnit(), IdempotencyKey: request.GetIdempotencyKey(),
		ExecutionType: request.GetExecutionType(), LimitPrice: request.GetLimitPrice(),
	})
	if err != nil && order.ID == "" {
		return nil, mapPolymarketError(err)
	}
	return &polymarketv1.PlaceOrderResponse{Order: orderToProto(order)}, nil
}

func (s *PolymarketServer) CancelOrder(
	ctx context.Context,
	request *polymarketv1.CancelOrderRequest,
) (*polymarketv1.CancelOrderResponse, error) {
	result, err := s.service.CancelOrder(
		ctx, request.GetToken(), request.GetTradingAccountId(), request.GetOrderId(),
	)
	if err != nil {
		return nil, mapPolymarketError(err)
	}
	return &polymarketv1.CancelOrderResponse{
		OrderId: result.OrderID, Status: result.Status, Message: result.Message,
	}, nil
}

func (s *PolymarketServer) GetOrder(
	ctx context.Context,
	request *polymarketv1.GetOrderRequest,
) (*polymarketv1.GetOrderResponse, error) {
	order, err := s.service.GetOrder(ctx, request.GetToken(), request.GetOrderId())
	if err != nil {
		return nil, mapPolymarketError(err)
	}
	return &polymarketv1.GetOrderResponse{Order: orderToProto(order)}, nil
}

func (s *PolymarketServer) StreamAccountEvents(
	request *polymarketv1.StreamAccountEventsRequest,
	stream polymarketv1.PolymarketService_StreamAccountEventsServer,
) error {
	events, unsubscribe, err := s.service.SubscribeAccountEvents(
		stream.Context(), request.GetToken(), request.GetTradingAccountId(),
	)
	if err != nil {
		return mapPolymarketError(err)
	}
	defer unsubscribe()
	heartbeat := time.NewTicker(15 * time.Second)
	defer heartbeat.Stop()
	for {
		select {
		case <-stream.Context().Done():
			return nil
		case <-heartbeat.C:
			if err := stream.Send(&polymarketv1.AccountEvent{
				Type: "heartbeat", ServerTime: timestamppb.Now(),
			}); err != nil {
				return err
			}
		case event := <-events:
			if err := stream.Send(accountEventToProto(event)); err != nil {
				return err
			}
		}
	}
}

func marketToProto(market polymarket.Market) *polymarketv1.Market {
	return &polymarketv1.Market{
		Id: market.ID, ConditionId: market.ConditionID, Slug: market.Slug,
		Asset: market.Asset, Period: market.Period, Title: market.Title,
		WindowStart: timestamppb.New(market.WindowStart),
		WindowEnd:   timestamppb.New(market.WindowEnd),
		UpTokenId:   market.UpTokenID, DownTokenId: market.DownTokenID,
		TickSize: market.TickSize, NegativeRisk: market.NegativeRisk, Active: market.Active,
	}
}

func snapshotToProto(
	snapshot polymarket.Snapshot,
	series []polymarket.PricePoint,
	includeSeries bool,
) *polymarketv1.MarketSnapshot {
	points := make([]*polymarketv1.PricePoint, 0)
	if includeSeries {
		for _, point := range series {
			points = append(points, pricePointToProto(point))
		}
	}
	return &polymarketv1.MarketSnapshot{
		Market: marketToProto(snapshot.Market), OpenPrice: snapshot.OpenPrice,
		ChainlinkPrice: snapshot.ChainlinkPrice, UpBid: snapshot.UpBid,
		UpAsk: snapshot.UpAsk, DownBid: snapshot.DownBid, DownAsk: snapshot.DownAsk,
		PriceSeries: points, SourceUpdatedAt: timestamppb.New(snapshot.SourceUpdated),
		Stale: snapshot.Stale, Version: snapshot.Version,
	}
}

func pricePointToProto(point polymarket.PricePoint) *polymarketv1.PricePoint {
	return &polymarketv1.PricePoint{
		Timestamp: timestamppb.New(point.Timestamp), OpenPrice: point.OpenPrice,
		ChainlinkPrice: point.ChainlinkPrice,
	}
}

func openOrderToProto(order polymarket.OpenOrder) *polymarketv1.OpenOrder {
	return &polymarketv1.OpenOrder{
		Id: order.ID, ConditionId: order.ConditionID, TokenId: order.TokenID,
		MarketTitle: order.MarketTitle, Outcome: order.Outcome, Side: order.Side,
		Price: order.Price, OriginalSize: order.OriginalSize,
		MatchedSize: order.MatchedSize, RemainingSize: order.RemainingSize,
		Status: order.Status, OrderType: order.OrderType,
		CreatedAt: timestamppb.New(order.CreatedAt),
	}
}

func accountEventToProto(event polymarket.AccountEvent) *polymarketv1.AccountEvent {
	orders := make([]*polymarketv1.OpenOrder, 0, len(event.OpenOrders))
	for _, order := range event.OpenOrders {
		orders = append(orders, openOrderToProto(order))
	}
	result := &polymarketv1.AccountEvent{
		Type: event.Type, OpenOrders: orders,
		PortfolioChanged: event.PortfolioChanged, ServerTime: timestamppb.Now(),
	}
	if event.Order != nil {
		result.Order = openOrderToProto(*event.Order)
	}
	return result
}

func positionToProto(position polymarket.Position) *polymarketv1.Position {
	return &polymarketv1.Position{
		Id: position.ID, TradingAccountId: position.TradingAccountID,
		ConditionId: position.ConditionID, TokenId: position.TokenID,
		Market: position.Market, Outcome: position.Outcome, Size: position.Size,
		AveragePrice: position.AveragePrice, CurrentPrice: position.CurrentPrice,
		InitialValue: position.InitialValue, CurrentValue: position.CurrentValue,
		CashPnl: position.CashPnL, PercentPnl: position.PercentPnL,
		Redeemable:      position.Redeemable,
		SourceUpdatedAt: timestamppb.New(position.SourceUpdatedAt),
	}
}

func orderToProto(order polymarket.Order) *polymarketv1.Order {
	return &polymarketv1.Order{
		Id: order.ID, ClobOrderId: order.CLOBOrderID,
		TradingAccountId: order.TradingAccountID, MarketId: order.MarketID,
		TokenId: order.TokenID, Outcome: order.Outcome, Side: order.Side,
		RequestedAmount: order.RequestedAmount, AmountUnit: order.AmountUnit,
		FilledSize: order.FilledSize, AveragePrice: order.AveragePrice,
		Status: order.Status, ErrorCode: order.ErrorCode,
		CreatedAt: timestamppb.New(order.CreatedAt), UpdatedAt: timestamppb.New(order.UpdatedAt),
		ExecutionType: order.ExecutionType, LimitPrice: order.LimitPrice,
		ClobOrderType: order.CLOBOrderType, ErrorMessage: order.ErrorMessage,
	}
}

func mapPolymarketError(err error) error {
	if status.Code(err) != codes.Unknown {
		return err
	}
	var upstream *polymarket.CLOBError
	if errors.As(err, &upstream) {
		switch upstream.StatusCode {
		case 401, 403:
			return status.Error(codes.Unauthenticated, "polymarket credentials invalid; please rebind")
		case 425, 429:
			return status.Error(codes.ResourceExhausted, "polymarket upstream rate limited")
		}
	}
	switch {
	case errors.Is(err, context.DeadlineExceeded):
		return status.Error(codes.DeadlineExceeded, "polymarket upstream timed out")
	case errors.Is(err, polymarket.ErrInvalidArgument):
		return status.Error(codes.InvalidArgument, err.Error())
	case errors.Is(err, polymarket.ErrNotFound), errors.Is(err, pgx.ErrNoRows):
		return status.Error(codes.NotFound, "resource not found")
	case errors.Is(err, polymarket.ErrMarketClosed):
		return status.Error(codes.FailedPrecondition, err.Error())
	case errors.Is(err, polymarket.ErrCancelRejected):
		return status.Error(codes.FailedPrecondition, err.Error())
	case errors.Is(err, polymarket.ErrAccountStreamUnavailable):
		return status.Error(codes.Unavailable, err.Error())
	case errors.Is(err, polymarket.ErrInsufficientFunds),
		errors.Is(err, polymarket.ErrInsufficientPosition):
		return status.Error(codes.ResourceExhausted, err.Error())
	default:
		return status.Error(codes.Unavailable, "polymarket upstream unavailable")
	}
}

File: web/src/lib/api/polymarket.ts
import type {
  PolymarketAccountSummary,
  PolymarketMarket,
  PolymarketMarketSnapshot,
  PolymarketOpenOrder,
  PolymarketOrder,
  PolymarketPosition,
  PolymarketPricePoint,
} from "@/types/polymarket";
import { createIdempotencyKey } from "../idempotency-key";

function apiBaseUrl(): string {
  return (process.env.NEXT_PUBLIC_API_BASE_URL ?? "").replace(/\/+$/, "");
}

function url(path: string): string {
  return `${apiBaseUrl()}/api/v1/polymarket${path}`;
}

function decimal(value: unknown): number | null {
  if (value === "" || value == null) return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

async function errorMessage(response: Response, fallback: string): Promise<string> {
  const body = (await response.json().catch(() => ({}))) as { error?: string };
  return body.error || fallback;
}

function market(value: unknown): PolymarketMarket {
  const item = value as Record<string, unknown>;
  return {
    id: String(item.id ?? ""),
    conditionId: String(item.conditionId ?? ""),
    slug: String(item.slug ?? ""),
    asset: String(item.asset ?? "") as PolymarketMarket["asset"],
    period: String(item.period ?? "") as PolymarketMarket["period"],
    title: String(item.title ?? ""),
    windowStart: String(item.windowStart ?? ""),
    windowEnd: String(item.windowEnd ?? ""),
    upTokenId: String(item.upTokenId ?? ""),
    downTokenId: String(item.downTokenId ?? ""),
    tickSize: String(item.tickSize ?? ""),
    negativeRisk: Boolean(item.negativeRisk),
    active: Boolean(item.active),
  };
}

const MAX_CHART_POINTS = 300;

function downsamplePoints<T>(points: T[], maxPoints = MAX_CHART_POINTS): T[] {
  if (points.length <= maxPoints) return points;
  const step = (points.length - 1) / (maxPoints - 1);
  return Array.from({ length: maxPoints }, (_, index) => {
    const sourceIndex = Math.round(index * step);
    return points[Math.min(sourceIndex, points.length - 1)];
  });
}

export function mergeSnapshot(
  current: PolymarketMarketSnapshot | null,
  incoming: PolymarketMarketSnapshot,
): PolymarketMarketSnapshot {
  if (current && current.market.id !== incoming.market.id) {
    return {
      ...incoming,
      priceSeries: downsamplePoints(incoming.priceSeries),
    };
  }
  if (!current || incoming.priceSeries.length > 1) {
    return {
      ...incoming,
      priceSeries: downsamplePoints(incoming.priceSeries),
    };
  }
  if (incoming.priceSeries.length === 0) {
    return {
      ...current,
      ...incoming,
      priceSeries: current.priceSeries,
    };
  }
  const merged = [...current.priceSeries];
  for (const point of incoming.priceSeries) {
    const index = merged.findIndex((item) => item.timestamp === point.timestamp);
    if (index >= 0) merged[index] = point;
    else merged.push(point);
  }
  merged.sort(
    (left, right) =>
      new Date(left.timestamp).getTime() - new Date(right.timestamp).getTime(),
  );
  return {
    ...current,
    ...incoming,
    priceSeries: downsamplePoints(merged),
  };
}

export type PolymarketLiveQuote = Omit<PolymarketMarketSnapshot, "priceSeries">;

export function liveQuoteFromSnapshot(
  snapshot: PolymarketMarketSnapshot,
): PolymarketLiveQuote {
  const { priceSeries, ...quote } = snapshot;
  void priceSeries;
  return quote;
}

export function applySnapshotParts(
  currentSnapshot: PolymarketMarketSnapshot | null,
  currentChartPoints: PolymarketPricePoint[],
  incoming: PolymarketMarketSnapshot,
): { liveSnapshot: PolymarketMarketSnapshot; chartPoints: PolymarketPricePoint[] } {
  const merged = mergeSnapshot(currentSnapshot, incoming);
  const marketChanged = currentSnapshot?.market.id !== incoming.market.id;
  let chartPoints = currentChartPoints;
  if (marketChanged || incoming.priceSeries.length > 0) {
    chartPoints = merged.priceSeries;
  }
  return {
    liveSnapshot: { ...merged, priceSeries: chartPoints },
    chartPoints,
  };
}

export function mapSnapshot(value: unknown): PolymarketMarketSnapshot {
  const item = value as Record<string, unknown>;
  const points = Array.isArray(item.priceSeries) ? item.priceSeries : [];
  return {
    market: market(item.market),
    openPrice: decimal(item.openPrice),
    chainlinkPrice: decimal(item.chainlinkPrice),
    upBid: decimal(item.upBid),
    upAsk: decimal(item.upAsk),
    downBid: decimal(item.downBid),
    downAsk: decimal(item.downAsk),
    priceSeries: points.map((point) => {
      const source = point as Record<string, unknown>;
      return {
        timestamp: String(source.timestamp ?? ""),
        openPrice: decimal(source.openPrice),
        chainlinkPrice: decimal(source.chainlinkPrice),
      };
    }),
    sourceUpdatedAt: String(item.sourceUpdatedAt ?? ""),
    stale: Boolean(item.stale),
    version: String(item.version ?? ""),
  };
}

export async function fetchPolymarketMarkets(): Promise<PolymarketMarket[]> {
  const response = await fetch(url("/markets?active=true"), {
    headers: { Accept: "application/json" },
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error(await errorMessage(response, "无法加载 Polymarket 市场"));
  }
  const payload = (await response.json()) as { data?: unknown[] };
  return (payload.data ?? []).map(market);
}

export async function fetchPolymarketSnapshot(
  marketId: string,
  etag?: string,
): Promise<{ snapshot: PolymarketMarketSnapshot | null; etag: string }> {
  const response = await fetch(url(`/markets/${encodeURIComponent(marketId)}`), {
    headers: {
      Accept: "application/json",
      ...(etag ? { "If-None-Match": etag } : {}),
    },
    cache: "no-store",
  });
  if (response.status === 304) return { snapshot: null, etag: etag ?? "" };
  if (!response.ok) {
    throw new Error(await errorMessage(response, "无法加载市场快照"));
  }
  const payload = (await response.json()) as { data: unknown };
  return {
    snapshot: mapSnapshot(payload.data),
    etag: response.headers.get("ETag") ?? "",
  };
}

export function polymarketStreamUrl(marketId: string): string {
  return url(`/stream?marketId=${encodeURIComponent(marketId)}`);
}

export async function fetchPolymarketAccountSummary(
  accountId: number,
): Promise<PolymarketAccountSummary> {
  const response = await fetch(url(`/accounts/${accountId}/summary`), {
    credentials: "include",
    headers: { Accept: "application/json" },
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error(await errorMessage(response, "无法加载账户资产"));
  }
  const { data } = (await response.json()) as {
    data: Record<string, unknown>;
  };
  return {
    tradingAccountId: Number(data.tradingAccountId),
    accountName: String(data.accountName ?? ""),
    walletAddress: String(data.walletAddress ?? ""),
    availableBalance: decimal(data.availableBalance) ?? 0,
    positionValue: decimal(data.positionValue) ?? 0,
    totalAssets: decimal(data.totalAssets) ?? 0,
    sourceUpdatedAt: String(data.sourceUpdatedAt ?? ""),
    stale: Boolean(data.stale),
    bindingStatus: String(data.bindingStatus ?? ""),
  };
}

export async function fetchPolymarketPositions(
  accountId: number,
): Promise<{ positions: PolymarketPosition[]; stale: boolean }> {
  const response = await fetch(url(`/accounts/${accountId}/positions`), {
    credentials: "include",
    headers: { Accept: "application/json" },
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error(await errorMessage(response, "无法加载持仓"));
  }
  const payload = (await response.json()) as {
    data?: Record<string, unknown>[];
    stale?: boolean;
  };
  return {
    positions: (payload.data ?? []).map((item) => ({
      id: String(item.id ?? item.tokenId ?? ""),
      tradingAccountId: Number(item.tradingAccountId),
      conditionId: String(item.conditionId ?? ""),
      tokenId: String(item.tokenId ?? ""),
      market: String(item.market ?? ""),
      outcome: String(item.outcome ?? ""),
      size: decimal(item.size) ?? 0,
      averagePrice: decimal(item.averagePrice) ?? 0,
      currentPrice: decimal(item.currentPrice) ?? 0,
      initialValue: decimal(item.initialValue) ?? 0,
      currentValue: decimal(item.currentValue) ?? 0,
      cashPnl: decimal(item.cashPnl) ?? 0,
      percentPnl: decimal(item.percentPnl) ?? 0,
      redeemable: Boolean(item.redeemable),
      sourceUpdatedAt: String(item.sourceUpdatedAt ?? ""),
    })),
    stale: Boolean(payload.stale),
  };
}

export function mapPolymarketOpenOrder(value: unknown): PolymarketOpenOrder {
  const item = value as Record<string, unknown>;
  return {
    id: String(item.id ?? ""),
    tokenId: String(item.tokenId ?? ""),
    conditionId: String(item.conditionId ?? ""),
    marketTitle: String(item.marketTitle ?? ""),
    outcome: String(item.outcome ?? ""),
    side: String(item.side ?? "") as PolymarketOpenOrder["side"],
    price: decimal(item.price) ?? 0,
    originalSize: decimal(item.originalSize) ?? 0,
    matchedSize: decimal(item.matchedSize) ?? 0,
    remainingSize: decimal(item.remainingSize) ?? 0,
    status: String(item.status ?? ""),
    orderType: String(item.orderType ?? ""),
    createdAt: String(item.createdAt ?? ""),
  };
}

export interface PolymarketAccountEvent {
  type: "snapshot" | "order" | "portfolio" | "heartbeat";
  order: PolymarketOpenOrder | null;
  openOrders: PolymarketOpenOrder[];
  portfolioChanged: boolean;
}

export function mapPolymarketAccountEvent(value: unknown): PolymarketAccountEvent {
  const item = value as Record<string, unknown>;
  return {
    type: String(item.type ?? "heartbeat") as PolymarketAccountEvent["type"],
    order: item.order ? mapPolymarketOpenOrder(item.order) : null,
    openOrders: Array.isArray(item.openOrders)
      ? item.openOrders.map(mapPolymarketOpenOrder)
      : [],
    portfolioChanged: Boolean(item.portfolioChanged),
  };
}

export function polymarketAccountStreamUrl(accountId: number): string {
  return url(`/accounts/${accountId}/events`);
}

export async function fetchPolymarketOpenOrders(
  accountId: number,
): Promise<{ openOrders: PolymarketOpenOrder[]; stale: boolean }> {
  const response = await fetch(url(`/accounts/${accountId}/open-orders`), {
    credentials: "include",
    headers: { Accept: "application/json" },
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error(await errorMessage(response, "无法加载挂单"));
  }
  const payload = (await response.json()) as {
    data?: unknown[];
    stale?: boolean;
  };
  return {
    openOrders: (payload.data ?? []).map(mapPolymarketOpenOrder),
    stale: Boolean(payload.stale),
  };
}

export async function cancelPolymarketOrder(
  accountId: number,
  orderId: string,
): Promise<{ orderId: string; status: string; message: string }> {
  const response = await fetch(
    url(`/accounts/${accountId}/orders/${encodeURIComponent(orderId)}`),
    {
      method: "DELETE",
      credentials: "include",
      headers: { Accept: "application/json" },
      cache: "no-store",
    },
  );
  if (!response.ok) {
    throw new Error(await errorMessage(response, "撤单失败"));
  }
  const payload = (await response.json()) as {
    data?: { orderId?: unknown; status?: unknown; message?: unknown };
  };
  return {
    orderId: String(payload.data?.orderId ?? orderId),
    status: String(payload.data?.status ?? "canceled"),
    message: String(payload.data?.message ?? ""),
  };
}

export async function placePolymarketOrder(input: {
  tradingAccountId: number;
  marketId: string;
  outcome: "up" | "down";
  side: "buy" | "sell";
  amount: string;
  amountUnit: "usd" | "shares";
  executionType?: "book" | "limit";
  limitPrice?: string;
}): Promise<PolymarketOrder> {
  const idempotencyKey = createIdempotencyKey();
  const response = await fetch(url("/orders"), {
    method: "POST",
    credentials: "include",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      "Idempotency-Key": idempotencyKey,
    },
    body: JSON.stringify({ ...input, idempotencyKey }),
  });
  if (!response.ok) {
    throw new Error(await errorMessage(response, "订单提交失败"));
  }
  const { data } = (await response.json()) as {
    data: Record<string, unknown>;
  };
  return {
    id: String(data.id ?? ""),
    clobOrderId: String(data.clobOrderId ?? ""),
    tradingAccountId: Number(data.tradingAccountId),
    marketId: String(data.marketId ?? ""),
    tokenId: String(data.tokenId ?? ""),
    outcome: String(data.outcome) as "up" | "down",
    side: String(data.side) as "buy" | "sell",
    requestedAmount: decimal(data.requestedAmount) ?? 0,
    amountUnit: String(data.amountUnit) as "usd" | "shares",
    executionType: (String(data.executionType || "book") as "book" | "limit"),
    limitPrice: decimal(data.limitPrice),
    clobOrderType: (String(data.clobOrderType || "FAK") as "FAK" | "GTC"),
    filledSize: decimal(data.filledSize) ?? 0,
    averagePrice: decimal(data.averagePrice) ?? 0,
    status: String(data.status ?? ""),
    errorCode: String(data.errorCode ?? ""),
    errorMessage: String(data.errorMessage ?? ""),
    createdAt: String(data.createdAt ?? ""),
    updatedAt: String(data.updatedAt ?? ""),
  };
}
